File: leetcode/graph/bfs/0994_rotting_oranges.cpp
// LC 994. Rotting Oranges | Medium
#include <vector>
#include <queue>
using namespace std;

class Solution {
public:
    int orangesRotting(vector<vector<int>>& grid) {
        int m = grid.size(), n = grid[0].size(), fresh = 0, time = 0;
        queue<pair<int,int>> q;
        for (int i = 0; i < m; i++)
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == 2) q.push({i,j});
                else if (grid[i][j] == 1) fresh++;
            }
        int d[] = {0,1,0,-1,0};
        while (!q.empty() && fresh) {
            time++; int sz = q.size();
            while (sz--) {
                auto [i,j] = q.front(); q.pop();
                for (int x = 0; x < 4; x++) {
                    int ni = i+d[x], nj = j+d[x+1];
                    if (ni >= 0 && ni < m && nj >= 0 && nj < n && grid[ni][nj] == 1) {
                        grid[ni][nj] = 2; fresh--; q.push({ni,nj});
                    }
                }
            }
        }
        return fresh == 0 ? time : -1;
    }
};

File: leetcode/graph/bfs/0994_rotting_oranges.py
# LC 994. Rotting Oranges | Medium
from collections import deque

class Solution:
    def orangesRotting(self, grid):
        m, n = len(grid), len(grid[0]); q = deque(); fresh = 0
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 2: q.append((i, j))
                elif grid[i][j] == 1: fresh += 1
        time = 0
        while q and fresh:
            time += 1
            for _ in range(len(q)):
                i, j = q.popleft()
                for di, dj in [(0,1),(0,-1),(1,0),(-1,0)]:
                    ni, nj = i+di, j+dj
                    if 0 <= ni < m and 0 <= nj < n and grid[ni][nj] == 1:
                        grid[ni][nj] = 2; fresh -= 1; q.append((ni, nj))
        return time if fresh == 0 else -1
